### PR TITLE
Added (homotopy) Pushouts and Cocones, and some features/examples of said.

### PR DESCRIPTION
**Pull request by jake-87**
# Description

Added (homotopy) Pushouts and Cocones, and some features/examples of said.

This PR is almost certainly not perfect - I have tried my best to match the existing code and similar, and likewise have tried my best to write good quality prose, but I'm leaving this a WIP until I'm confident in the quality of work.

In the meantime though, review/critique is much appreciated.

## Checklist

Before submitting a merge request, please check the items below:

- [x] I've read [the contributing guidelines](https://github.com/plt-amy/1lab/blob/main/CONTRIBUTING.md).
- [x] The imports of new modules have been sorted with `support/sort-imports.hs` (or `nix run --experimental-features nix-command -f . sort-imports`).
- [x] All new code blocks have "agda" as their language.

If your change affects many files without adding substantial content, and
you don't want your name to appear on those pages (for example, treewide
refactorings or reformattings), start the commit message and PR title with `chore:`.

### CODE REVIEW

**Comment by Lavenza** [comment]
[Pull request preview](https://preview.1lab.dev/418)

File: src/Homotopy/Pushout.lagda.md [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  iso-pf .rinv (inl x) = refl
+  iso-pf .rinv (inr x) = refl
+  iso-pf .rinv (commutes c i) = refl
+  iso-pf .linv N = refl
+  iso-pf .linv S = refl
+  iso-pf .linv (merid x i) = refl
+```
+</details>
+
+## The Universal property of a pushout, via Cocones
```
Comment (TOTBWF) [inline src/Homotopy/Pushout.lagda.md:150]: Universal should be lowercase here.

File: src/Homotopy/Pushout.lagda.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+      ((c : C) → i (f c) ≡ j (g c))
+```
+
+We can then show that the canonical `Cocone`{.Agda} consisting of a `Pushout`{.Agda}
+is the universal `Cocone`{.Agda}.
+
+```agda
+Pushout→E≡CoconeE : ∀ {A B C E f g} →
+                    (Pushout C A f B g → E) ≡ (Cocone f g E)
+Pushout→E≡CoconeE = ua ( Pushout→Cocone , is-iso→is-equiv iso-pc ) where
```
Comment (TOTBWF) [inline src/Homotopy/Pushout.lagda.md:220]: We should probably remove redundant spaces here. Also, maybe this should be `Pushout-universal-cocone` or something?

**Review by TOTBWF — COMMENTED** [review]
Thanks for the contribution! Just a couple of minor formatting nits :)

File: src/Homotopy/Pushout.lagda.md (hunk 2 above)
Comment (jake-87) [reply src/Homotopy/Pushout.lagda.md]: Hm, which spaces are you referring to? The newline could be removed, but are you also referring to others?

**Comment by jake-87** [comment]
Sorry about that commit sleugh, I'm happy to squash them if wanted.

File: src/Homotopy/Pushout.lagda.md [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+\[\begin{tikzcd}
+  C && B \\
+  \\
+  A
+  \arrow["g", from=1-1, to=1-3]
+  \arrow["f"', from=1-1, to=3-1]
+\end{tikzcd}\]
+~~~
+
+The **Pushout** of this span is defined as the higher inductive type
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:28]:
```suggestion
The **pushout** of this span is defined as the higher inductive type
```

File: src/Homotopy/Pushout.lagda.md [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+```agda
+data Pushout {ℓ} (C : Type ℓ)
+                 (A : Type ℓ) (f : C → A)
+                 (B : Type ℓ) (g : C → B)
+                 : Type ℓ where
+```
+
+Two functions `inl`{.Agda} and `inr`{.Agda}, that "connect" from `A` and `B`:
+
+```agda
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:39]:
```suggestion
```

File: src/Homotopy/Pushout.lagda.md [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+  inl : A → Pushout C A f B g
+  inr : B → Pushout C A f B g
+```
+
+And, for every `c : C`, a path
+`inl (f c) ≡ inr (g c)`{.Agda}:
+
+```agda
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:47]:
```suggestion
```

File: src/Homotopy/Pushout.lagda.md [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+These combine to give the following:
+
+~~~{.quiver}
+\[\begin{tikzcd}
+	C && B \\
+	\\
+	A && {\rm{Pushout}}
+	\arrow["g", from=1-1, to=1-3]
+	\arrow["f"', from=1-1, to=3-1]
+	\arrow["{\rm{inl}}", from=1-3, to=3-3]
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:60]:
```suggestion
	\arrow["{\rm{inr}}", from=1-3, to=3-3]
```

File: src/Homotopy/Pushout.lagda.md [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+	\arrow["{\rm{commutes}}"{description}, shorten <=6pt, shorten >=6pt, Rightarrow, from=3-1, to=1-3]
+	\arrow["{\rm{inl}}"', from=3-1, to=3-3]
+\end{tikzcd}\]
+~~~
+
+## Suspensions as Pushouts
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:66]:
```suggestion
## Suspensions as pushouts
```

File: src/Homotopy/Pushout.lagda.md [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+First, recall the `Susp`{.Agda}, which serves to increase the
+connectedness of a space. Suspension can be expressed as a
+`Pushout`{.Agda} of, for some type `A`, the span `⊤ ← A → ⊤`{.Agda}:
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:70]:
```suggestion
The [[suspension]] of a type $A$ can be expressed as the `Pushout`{.Agda}
$\top \ot A \to \top$:
```

File: src/Homotopy/Pushout.lagda.md [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+
+~~~{.quiver}
+\[\begin{tikzcd}
+	C && \top \\
+	\\
+	\top && {\rm{Pushout}}
+	\arrow["{\rm{const\,tt}}", from=1-1, to=1-3]
+	\arrow["{\rm{const\,tt}}"', from=1-1, to=3-1]
+	\arrow["{\rm{inl}}", from=1-3, to=3-3]
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:79]:
```suggestion
	\arrow["{!}", from=1-1, to=1-3]
	\arrow["{!}"', from=1-1, to=3-1]
	\arrow["{\rm{inr}}", from=1-3, to=3-3]
```

File: src/Homotopy/Pushout.lagda.md [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+obvious that this is equivalent to suspension - because both of our
+non-path constructors have input `⊤`{.Agda}, they're indistiguishable
+from members of the pushout; therefore, we take the
+`inl`{.Agda} and `inr`{.Agda} to `N`{.Agda} and
+`S`{.Agda} respectively.
+Likewise, we take `commutes`{.Agda} to
+`merid`{.Agda}, as our functions `f`{.Agda} and `g`{.Agda} can only
+be `to-top`{.Agda}, so it reduces (with a suitable element of `C`{.Agda})
+from `∀ c → inl (f c) ≡ inr (g c)`{.Agda} to `inl tt ≡ inr tt`{.Agda},
+or equivalently `inl ≡ inr`{.Agda}.
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:105]:
```suggestion
`merid`{.Agda}.
```

File: src/Homotopy/Pushout.lagda.md [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+We therefore aim to show:
+
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:108]:
```suggestion
```

File: src/Homotopy/Pushout.lagda.md [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+```agda
+Susp≡Pushout-⊤←A→⊤ : ∀ {A} → Susp A ≡ Pushout A ⊤ (const tt) ⊤ (const tt)
+```
+
+The left and right functions are therefore trival:
+
+```agda
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:115]:
```suggestion
```

File: src/Homotopy/Pushout.lagda.md [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+  iso-pf .rinv (inl x) = refl
+  iso-pf .rinv (inr x) = refl
+  iso-pf .rinv (commutes c i) = refl
+  iso-pf .linv N = refl
+  iso-pf .linv S = refl
+  iso-pf .linv (merid x i) = refl
+```
+</details>
+
+## The universal property of a pushout, via Cocones
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:150]:
```suggestion
## The universal property of pushouts
```

File: src/Homotopy/Pushout.lagda.md [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+
+To formulate the universal property of a pushout, we first introduce the **Cocone**.
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:152]:
```suggestion
To formulate the universal property of a pushout, we first introduce **cocones**.
```

File: src/Homotopy/Pushout.lagda.md [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+```
+-->
+```agda
+  i : A → D
+  j : B → D
+``` 
+and a homotopy
+```agda
+  h : (c : C) → i (f c) ≡ j (g c)
+```
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:178]: These are unused in the code: replace them with inline math.

**Review by ncfavier — CHANGES_REQUESTED** [review]

**Comment by jake-87** [comment]
Thank you - I'll take that as a general comment that my writing style was too "informal", I suppose?

File: src/Homotopy/Pushout.lagda.md [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+~~~{.quiver}
+\[\begin{tikzcd}
+	A & C & B
+	\arrow["f"', from=1-2, to=1-1]
+	\arrow["g", from=1-2, to=1-3]
+\end{tikzcd}\]
+~~~
+
+consists of functions `i : A → D`{.Agda} & `j : B → D`{.Agda}, and a homotopy
+`h : (c : C) → i (f c) ≡ j (g c)`{.Agda}, forming:
```
Comment (ncfavier) [inline src/Homotopy/Pushout.lagda.md:143]:
```suggestion
consists of functions $i : A \to D$, $j : B \to D$, and a homotopy
$h : (c : C) \to i (f c) \is j (g c)$, forming:
```

**Comment by jake-87** [comment]
Hi @ncfavier - is there anything else you'd like me to change?

**Review by ncfavier — APPROVED** [review]

**Pull request merged**